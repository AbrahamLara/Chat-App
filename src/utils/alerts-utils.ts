export enum ErrorPayloadType {
  REGISTER = 'register',
  LOGIN = 'login',
}

/**
 * An alert message object.
 */
export interface AlertMessage {
  /**
   * The message to display.
   */
  message: string;
}

/**
 * Describes an alert banner that displays messages for the user to see on a page.
 */
export interface AlertBarOptions extends AlertMessage {
  /**
   * Determines if the banner should be visible.
   */
  visible: boolean;

  /**
   * The type of alert.
   */
  type: string;
}

/**
 * Describes an error message for an invalid form field value.
 */
export interface FormAlertMessage extends AlertMessage {
  /**
   * The form field the error message is for. This may not to be provided if a form error wasn't due to an invalid field
   * value.
   */
  field?: string;
}

/**
 * Describes the payload to return for a failed request.
 */
export interface ErrorPayload {
  /**
   * The error type.
   *
   * Note: This might not actually be needed.
   */
  type: ErrorPayloadType;

  /**
   * The error messages produced. Right now the backend only serves form errors.
   */
  errors: FormAlertMessage[];
}

export const DEFAULT_ALERT_BAR_OPTIONS: AlertBarOptions = {
  visible: false,
  type: '',
  message: '',
};

/**
 * Determines if the give value is an error response object sent from a server request.
 */
export function isFormErrorMessage(response: any) {
  return typeof response === 'object' && response.type && response.errors;
}

/**
 * Creates a generic message object to display as an alert.
 */
export function createAlertMessage(message: string): AlertMessage {
  return { message };
}
