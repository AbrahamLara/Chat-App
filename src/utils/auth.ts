export interface RegexRule {
  regex: RegExp;

  rule: string;

  passed?: boolean;
}

/**
 * Password regex also used in the backend.
 * [PASSWORD_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L22-L41)
 */
export const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*\-_]).{8,}$/;

/**
 * Email regex also used in the backend.
 * [EMAIL_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L43-L48)
 */
const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const EMAIL_FORM_RULES = [
  (email: string) => !!email || 'Email is required',
  (email: string) => EMAIL_REGEX.test(email) || 'Email must be valid.',
];

export const PASSWORD_FORM_RULES = [
  (password: string) => !!password || 'Password is required',
  (password: string) =>
    PASSWORD_REGEX.test(password) || 'Password must be valid.',
];

export const PASSWORD_REGEX_RULES: RegexRule[] = [
  { rule: '8 characters minimum', regex: /.{8,}/ },
  { rule: '1 uppercase character', regex: /[A-Z]/ },
  { rule: '1 lowercase character', regex: /[a-z]/ },
  { rule: '1 number', regex: /[0-9]/ },
  {
    rule: '1 special character: !, @, #, $, %, ^, &, *, -, _',
    regex: /[!@#$%^&*\-_]/,
  },
];
