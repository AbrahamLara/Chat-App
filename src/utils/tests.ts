/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 */
function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

export { resolveVuetifyAppDataWarning };
