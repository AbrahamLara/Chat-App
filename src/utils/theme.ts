// Themes as string enums since the app state theme should only be either of these string values.
export enum THEME {
  DARK = 'Dark',
  LIGHT = 'Light'
}

// The application theme background color hexes.
export const ThemeHexes = {
  [THEME.DARK]: '#121212',
  [THEME.LIGHT]: '#fff'
};
