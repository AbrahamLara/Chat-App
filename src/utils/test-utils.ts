import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}
