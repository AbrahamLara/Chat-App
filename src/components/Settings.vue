<template>
  <v-menu v-model="showMenu" rounded offset-y nudge-left="60px">
    <template v-slot:activator="{ attrs, on }">
      <v-btn
        elevation="0"
        v-bind="attrs"
        v-on="on"
        style="font-size: 16px"
        icon
      >
        <v-icon>mdi-cog-outline</v-icon>
      </v-btn>
    </template>
    <v-list>
      <v-list-item-group v-model="selectedThemeItem">
        <v-list-item
          v-for="(theme, index) in themes"
          :key="theme.title"
          active-class="primary--text"
          link
          @click="handleTheme(theme.title)"
          :disabled="selectedThemeItem === index"
        >
          <v-list-item-title
            class="text-capitalize"
            v-text="theme.title"
          ></v-list-item-title>
          <v-list-item-icon>
            <v-icon v-text="theme.icon"></v-icon>
          </v-list-item-icon>
        </v-list-item>
        <v-list-item
          v-if="matchMediaIsSupported"
          active-class="primary--text"
          link
          @click="addSystemThemeListener"
          :disabled="selectedThemeItem === 2"
        >
          <v-list-item-title class="text-capitalize">System</v-list-item-title>
          <v-list-item-icon>
            <v-icon>mdi-desktop-tower-monitor</v-icon>
          </v-list-item-icon>
        </v-list-item>
      </v-list-item-group>
    </v-list>
  </v-menu>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { THEME } from '@/utils/theme';
import { Action } from '@/utils/decorators';
import { ActionMethod } from 'vuex';
import { SET_THEME } from '@/store/actions';

// Represents the theme item to display in a list.
interface ThemeListItem {
  // The list item title to represent the theme.
  title: THEME;

  // The list item icon to represent the theme.
  icon: string;
}

@Component({ name: 'settings-btn' })
export default class Settings extends Vue {
  @Action(SET_THEME) updateAppTheme!: ActionMethod;
  // The themes to display in the settings dropdown
  readonly themes: ThemeListItem[] = [
    { title: THEME.LIGHT, icon: 'mdi-white-balance-sunny' },
    { title: THEME.DARK, icon: 'mdi-weather-night' }
  ];
  // This variable solely exists for testing purposes.
  showMenu = false;

  /**
   * The selected theme item on the settings list.
   *
   * 0 = Light
   *
   * 1 = Dark
   *
   * 2 = System
   */
  selectedThemeItem = 0;

  // The MediaQueryList object that will be helpful in changing the theme of the application to match the system
  // theme.
  matchMedia: MediaQueryList | null = null;

  // A computed property that determines if matchMedia is supported in the current browser.
  get matchMediaIsSupported() {
    return Boolean(window.matchMedia);
  }

  // Handles the theme item selected.
  handleTheme(theme: THEME) {
    this.removeSystemThemeListener();
    this.updateAppTheme(theme);
  }

  /**
   * This function is for listening to system theme changes based on a media query event.
   *
   * @param event A MediaQueryListEvent object.
   */
  handleSystemTheme(event: MediaQueryListEvent) {
    const theme = event.matches ? THEME.DARK : THEME.LIGHT;
    this.updateAppTheme(theme);
  }

  // Adds a change listener to automatically update the application's theme based on the system theme.
  addSystemThemeListener() {
    if (this.matchMedia === null) {
      // The MediaQueryList object that helps determine if the system theme is dark mode.
      this.matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
      // Set the theme based on whether or not the media query matched.
      const theme = this.matchMedia.matches ? THEME.DARK : THEME.LIGHT;
      this.updateAppTheme(theme);
      // Add the change listener that will determine application's theme based on the system theme.
      this.matchMedia.addEventListener('change', this.handleSystemTheme, false);
    }
  }

  // Removes the change listener that updates the application's theme based on the system theme.
  removeSystemThemeListener() {
    if (this.matchMedia !== null) {
      this.matchMedia.removeEventListener(
        'change',
        this.handleSystemTheme,
        false
      );
      this.matchMedia = null;
    }
  }
}
</script>
