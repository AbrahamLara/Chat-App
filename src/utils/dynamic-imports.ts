// This file contains dynamic imports of components to reduce bundle size when building project with webpack.

const getAbout = () =>
  import(/* webpackChunkName: "About" */ '@/components/About.vue');

const getAppBar = () =>
  import(/* webpackChunkName: "AppBar" */ '@/components/AppBar.vue');

const getAuthPage = () =>
  import(/* webpackChunkName: "AuthPage"  */ '@/views/AuthPage.vue');

const getFormCard = () =>
  import(/* webpackChunkName: "FormCard" */ '@/components/FormCard.vue');

const getNotFound = () =>
  import(/* webpackChunkName: "NotFound" */ '@/views/NotFound.vue');

const getRegexRulesTester = () =>
  import(
    /* webpackChunkName: "RegexRulesTester" */ '@/components/RegexRulesTester.vue'
  );

const getSettings = () =>
  import(/* webpackChunkName: "Settings" */ '@/components/Settings.vue');

const getSignIn = () =>
  import(/* webpackChunkName: "SignIn" */ '@/components/SignIn.vue');

const getSignUp = () =>
  import(/* webpackChunkName: "SignUp" */ '@/components/SignUp.vue');

const getSplash = () =>
  import(/* webpackChunkName: "Splash" */ '@/views/Splash.vue');

export {
  getAbout,
  getAppBar,
  getAuthPage,
  getFormCard,
  getNotFound,
  getRegexRulesTester,
  getSettings,
  getSignIn,
  getSignUp,
  getSplash,
};
