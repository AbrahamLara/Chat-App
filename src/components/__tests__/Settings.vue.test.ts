import Vue from 'vue';
import Vuetify from 'vuetify';
import Settings from '@/components/Settings.vue';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import { resolveVuetifyAppDataWarning } from '@/utils/tests';

resolveVuetifyAppDataWarning();

const localVue = createLocalVue();
Vue.use(Vuetify);

describe('Settings.vue.test.ts', () => {
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const vuetify = new Vuetify();
    wrapper = mount(Settings, { localVue, vuetify });
  });

  it('renders settings button correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('shows menu when button is clicked', async () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    const mockClickEvent = { stopPropagation: () => {} };
    jest.spyOn(mockClickEvent, 'stopPropagation');
    // Find the menu component.
    const menu = wrapper.find('.v-menu');
    expect(menu.exists()).toBeTruthy();
    // The menu should not be shown.
    expect(wrapper.vm.showMenu).toBeFalsy();
    // Find the button and click it to show the menu.
    const button = menu.find('.v-btn');
    button.vm.$emit('click', mockClickEvent);
    await wrapper.vm.$nextTick();
    // The click event calls stopPropagation.
    expect(mockClickEvent.stopPropagation).toHaveBeenCalledTimes(1);
    // The menu element should've updated to reflect it being shown.
    const updatedMenu = wrapper.find('.v-menu');
    expect(updatedMenu.element).toMatchSnapshot();
    // The menu should be shown now.
    expect(wrapper.vm.showMenu).toBeTruthy();
  });

  it('attaches and removes system theme listener correctly', () => {
    // Mock the handleSystemTheme method since we aren't testing vuex.
    wrapper.vm.updateAppTheme = jest.fn();
    // The matchMedia class property should be null by default.
    expect(wrapper.vm.matchMedia).toBeNull();
    // Create a mock MediaQueryList object with event listener methods.
    const mockMediaQueryList = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    } as any;
    // Mock window.matchMedia.
    window.matchMedia = jest.fn(() => mockMediaQueryList);

    wrapper.vm.addSystemThemeListener();
    // window.matchMedia should've been called.
    expect(window.matchMedia).toHaveBeenCalled();
    // The method to add an event listener to the match media object should've been called.
    expect(mockMediaQueryList.addEventListener).toHaveBeenCalled();
    // The updateAppTheme method should've been called as a result of calling addSystemThemeListener.
    expect(wrapper.vm.updateAppTheme).toHaveBeenCalled();
    // The matchMedia class property should no longer be null.
    expect(wrapper.vm.matchMedia).not.toBeNull();

    wrapper.vm.removeSystemThemeListener();
    // The method to remove an event listener from the match media object should've been called.
    expect(mockMediaQueryList.removeEventListener).toHaveBeenCalled();
    // The matchMedia class property should be null as a result of removing the listener.
    expect(wrapper.vm.matchMedia).toBeNull();
  });
});
