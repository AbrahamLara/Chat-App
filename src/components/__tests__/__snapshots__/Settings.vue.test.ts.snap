// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings.vue.test.ts renders settings button correctly 1`] = `
<div
  class="v-menu"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
    role="button"
    style="font-size: 16px;"
    type="button"
  >
    <span
      class="v-btn__content"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate mdi mdi-cog-outline theme--light"
      />
    </span>
  </button>
  <!---->
</div>
`;

exports[`Settings.vue.test.ts shows menu when button is clicked 1`] = `
<div
  class="v-menu"
>
  <button
    aria-expanded="true"
    aria-haspopup="true"
    class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
    role="button"
    style="font-size: 16px;"
    type="button"
  >
    <span
      class="v-btn__content"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate mdi mdi-cog-outline theme--light"
      />
    </span>
  </button>
  <transition-stub
    name="v-menu-transition"
  />
</div>
`;
